File: Leetcode/sort_and_search/278_FirstBadVersion/Solution.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 0, end = n;
        while(start<=end){
            int mid = start + (end-start)/2;
            if(!isBadVersion(mid)){
                start = mid+1;
            }else if(mid==0 || !isBadVersion(mid-1) ){
                return mid;
            }else{
                end = mid-1;
            }
        }
        return -1;
    }
}

File: Leetcode/sort_and_search/278_FirstBadVersion/Solution.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        start, end = 0, n
        while start <= end:
            mid = (start+end)//2
            if not isBadVersion(mid):
                start = mid+1
            elif mid == 0 or not isBadVersion(mid-1):
                return mid
            else:
                end = mid-1
        return -1
